File: prep/implementation/main_iterations/iteration6/AdjacencySetBFS.java
import java.util.*;

public class AdjacencySetBFS {
  static interface Graph {
    void addEdge(int from, int to);
    List<Integer> getAdjacentVertices(int v);
    int size();
  }

  static class AdjacencySet implements Graph {
    static class Vertex {
      Set<Integer> vs;
      Vertex() { this.vs = new HashSet<>(); }

      void addEdge(int v) { this.vs.add(v); }
      List<Integer> getAdjacentVertices() {
        List<Integer> ns = new ArrayList<>();
        for (int v : this.vs) {
          ns.add(v);
        }

        return ns;
      }
    }

    private List<Vertex> vertices;
    private int size;

    AdjacencySet(int size) {
      this.size = size;
      this.vertices = new ArrayList<>(size);

      for (int i = 0; i < size; i++) {
        this.vertices.add(new Vertex());
      }
    }

    @Override
    public void addEdge(int from, int to) {
      if (from < 0 || from >= this.size || to < 0 || to >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      this.vertices.get(from).addEdge(to);
    }

    @Override
    public List<Integer> getAdjacentVertices(int v) {
      if (v < 0 || v >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      return this.vertices.get(v).getAdjacentVertices();
    }

    @Override
    public int size() {
      return this.size;
    }
  }

  public static void bfs(Graph g) {
    boolean[] visited = new boolean[g.size()];
    for (int i = 0; i < g.size(); i++) {
      if (!visited[i]) {
        bfs(g, visited, i);
      }
    }
    System.out.println();
  }

  private static void bfs(Graph g, boolean[] visited, int currVertex) {
    Queue<Integer> q = new ArrayDeque<>();
    q.add(currVertex);

    while (!q.isEmpty()) {
      int v = q.poll();

      visited[v] = true;
      System.out.printf("%d ", v);

      for (int neighbour : g.getAdjacentVertices(v)) {
        if (!visited[neighbour]) {
          q.add(neighbour);
        }
      }
    }
  }

  public static void bfsRec(Graph g) {
    boolean[] visited = new boolean[g.size()];
    for (int i = 0; i < g.size(); i++) {
      if (!visited[i]) {
        Queue<Integer> q = new ArrayDeque<>();
        q.add(i);
        bfsRec(g, visited, q);
      }
    }
    System.out.println();
  }

  private static void bfsRec(Graph g, boolean[] visited, Queue<Integer> q) {
    if (q.isEmpty()) {
      return;
    }

    int v = q.poll();
    visited[v] = true;
    System.out.printf("%d ", v);

    for (int neighbour : g.getAdjacentVertices(v)) {
      if (!visited[neighbour]) {
        q.add(neighbour);
      }
    }

    bfsRec(g, visited, q);
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int m = in.nextInt();

      Graph g = new AdjacencySet(n);
      for (int i = 0; i < m; i++) {
        int from = in.nextInt();
        int to = in.nextInt();

        g.addEdge(from, to);
        g.addEdge(to, from);
      }

      bfs(g);
      bfsRec(g);
    }
  }
}

File: lc/coreprep/implementation/main_iterations/iteration_cpp/adjacency_set_bfs.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <set>
#include <vector>

using namespace std;

struct vertex {
  set<int> vs;

  void add_edge(int v) { this->vs.insert(v); }

  const vector<int> get_adjacent_vertices() const {
    vector<int> neighbours;

    for (int v : this->vs) {
      neighbours.push_back(v);
    }

    return neighbours;
  }
};

class graph {
private:
  vector<vertex> vertices;
  int sz;

public:
  graph(int sz) : sz(sz) {
    for (int i = 0; i < sz; i++) {
      vertex v;
      this->vertices.emplace_back(v);
    }
  }

  void add_edge(int from, int to) {
    if (from < 0 || from >= this->sz || to < 0 || to >= this->sz) {
      throw "illegal vertex";
    }

    this->vertices[from].add_edge(to);
  }

  const vector<int> get_adjacent_vertices(int v) const {
    if (v < 0 || v >= this->sz) {
      throw "illegal vertex";
    }

    return this->vertices[v].get_adjacent_vertices();
  }

  int size() const { return this->sz; }
};

void bfs(graph &g, vector<bool> &visited, int curr_vertex) {
  queue<int> q;
  q.push(curr_vertex);

  while (!q.empty()) {
    int v = q.front();
    q.pop();

    visited[v] = true;
    cout << v << " ";

    for (int neighbour : g.get_adjacent_vertices(v)) {
      if (!visited[neighbour]) {
        q.push(neighbour);
      }
    }
  }
}

void bfs(graph &g) {
  vector<bool> visited(g.size());
  fill(visited.begin(), visited.end(), false);

  for (int i = 0; i < g.size(); i++) {
    if (!visited[i]) {
      bfs(g, visited, i);
    }
  }
  cout << "\n";
}

void bfs_rec(graph &g, vector<bool> &visited, queue<int> &q) {
  if (q.empty()) {
    return;
  }

  int v = q.front();
  q.pop();

  visited[v] = true;
  cout << v << " ";

  for (int neighbour : g.get_adjacent_vertices(v)) {
    if (!visited[neighbour]) {
      q.push(neighbour);
    }
  }

  bfs_rec(g, visited, q);
}

void bfs_rec(graph &g) {
  vector<bool> visited(g.size());
  fill(visited.begin(), visited.end(), false);

  for (int i = 0; i < g.size(); i++) {
    if (!visited[i]) {
      queue<int> q;
      q.push(i);
      bfs_rec(g, visited, q);
    }
  }
  cout << "\n";
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, from, to;
  cin >> n >> m;

  graph g(n);
  for (int i = 0; i < m; i++) {
    cin >> from >> to;
    g.add_edge(from, to);
    g.add_edge(to, from);
  }

  bfs(g);
  bfs_rec(g);

  return 0;
}